### DISCUSSION TITLE
Are vault plugins supposed to work with openbao?

### DISCUSSION THREAD

**Discussion by DrDaveD** [openbao/openbao#317] [Q&A] [answered]
I'm trying to load a couple of plugins compiled for vault with the current openbao main branch and I get errors like:

```
Error enabling: Error making API request.

URL: POST http://127.0.0.1:8202/v1/sys/mounts/secret/oauth
Code: 400. Errors:

* invalid backend version: 2 errors occurred:
        * Unrecognized remote plugin message:
This usually means
  the plugin was not compiled for this architecture,
  the plugin is missing dynamic-link libraries necessary to run,
  the plugin is not executable by this process due to file permissions, or
  the plugin failed to negotiate the initial go-plugin protocol handshake
...
```

Is it supposed to work at this point?  Do the plugins need to be modified to use the openbao libraries and recompiled?

**Comment by cipherboy** [comment]
@DrDaveD Interesting! I had wanted this to work, and would've thought it should've, but apparently I broke something. Do you mind modifying the code to see if you can get which message is unrecognized?

I'm also curious about versions. I'd expect it to work if you chose an upstream SDK corresponding to Vault 1.14.x release, but not necessarily if you chose later ones... Unless I broke it due to removing too much eventing code....?

**Reply by DrDaveD** [reply]
I don't mind modifying the code.  Where exactly do you suggest I start with that?

The plugins that I attempted to load are
- https://github.com/puppetlabs/vault-plugin-secrets-oauthapp v3.1.1
- https://github.com/42wim/vault-plugin-auth-ssh v0.3.1

According to their go.mod files, they're using hashicorp/vault/sdk v0.6.1-0.20230302210543-38f40f637f4f and v0.10.2 respectively.  So, quite old.

**Reply by cipherboy** [reply]
Interesting! 

I was hoping this one could be made a little more clear:

> Unrecognized remote plugin message:

Any chance you could target the same SDK version but use a `replace` directive on that version to make clear _which_ plugin message it is?

In Vault 1.13 (iirc?) plugin v5 framework was introduced. I think we retained v4 and v5 compatibility, but maybe dropped earlier formats, so maybe also indicating which plugin message format is used would be good to have too?

**Reply by DrDaveD** [reply]
So far I learned that the sdk tags are independent of the vault version.  The latest sdk version is v0.11.1, so at least the second plugin was not far behind.  I tried rebuilding it with v0.11.1 and saw no difference.

**Reply by cipherboy** [reply]
Weird... So oauth uses:

https://github.com/puppetlabs/vault-plugin-secrets-oauthapp/blob/ed4005c19ac7ab9f6f152379a338168021300a21/go.mod#L178

> ```
> github.com/hashicorp/go-plugin v1.4.5 // indirect
> ```

which is a similar version to what we have in our deps...

https://github.com/openbao/openbao/blob/a5e299de79465b922076f415e8945efaf0c9eb91/go.mod#L61

(but different point release...).

However, while the location of the message changed, it is still emitted on a line with less than 4 parts (delineated by `|`)... However, all three versions in play (ssh uses a different one) should actually emit the line from the remote peer:

> ```go
> const unrecognizedRemotePluginMessage = `Unrecognized remote plugin message: %s
> ... snip ...
> ```

Which means, the remote peer's message is... empty?

Are there any log messages from the server side when this plugin is mounted?

**Reply by DrDaveD** [reply]
Yes there are debug log messages in /var/log/messages:
```
May  3 17:40:11 fermicloud732 bao[2108738]: 2024-05-03T17:40:11.984-0500 [DEBUG] auth.ssh.auth_ssh_e5bf4da6.ssh: waiting for RPC address: metadata=true path=/usr/libexec/htvault-config/plugins/vault-plugin-auth-ssh.sh
May  3 17:40:11 fermicloud732 bao[2108738]: 2024-05-03T17:40:11.993-0500 [ERROR] auth.ssh.auth_ssh_e5bf4da6.ssh.vault-plugin-auth-ssh.sh: plugin tls init: metadata=true error="error parsing wrapping token: go-jose/go-jose: compact JWS format must have three parts" timestamp=2024-05-03T17:40:11.993-0500
May  3 17:40:11 fermicloud732 bao[2108738]: 2024-05-03T17:40:11.994-0500 [INFO] auth.ssh.auth_ssh_e5bf4da6.ssh: plugin process exited: metadata=true path=/usr/libexec/htvault-config/plugins/vault-plugin-auth-ssh.sh pid=2108893
May  3 17:40:11 fermicloud732 bao[2108738]: 2024-05-03T17:40:11.994-0500 [ERROR] secrets.system.system_61f5ed54: error occurred during enable credential: path=ssh/
May  3 17:40:11 fermicloud732 bao[2108738]:  error=
May  3 17:40:11 fermicloud732 bao[2108738]:  | invalid backend version: 2 errors occurred:
May  3 17:40:11 fermicloud732 bao[2108738]:  | \t* Unrecognized remote plugin message:
```
I think the key is the tls error

**Reply by DrDaveD** [reply]
If I try to execute one of the plugins directly with environment `VAULT_BACKEND_PLUGIN=6669da05-b1c8-4f49-97d9-c8e5bed98e20` (as found at the end of [sdk/plugin/serve.go](https://github.com/openbao/openbao/blob/main/sdk/plugin/serve.go#L165) then I get as output something that looks very similar:
```
{"@level":"error","@message":"plugin tls init","@timestamp":"2024-05-03T21:40:50.000249-05:00","error":"error parsing wrapping token: square/go-jose: compact JWS format must have three parts"}
```
So there's something wrong with the tls wrapper.

**Reply by DrDaveD** [reply]
Aha!  I had a plugin wrapper dump the environment and I saw 3 variables with a BAO prefix:
```
BAO_PLUGIN_AUTOMTLS_ENABLED=true
BAO_PLUGIN_METADATA_MODE=false
BAO_VERSION=2.0.0
```
If I copy those to corresponding VAULT variables, I can get the plugins to start.

**Reply by cipherboy** [reply]
Aha, sorry. This makes a lot more sense now. I'll fix this shortly!

**Comment by DrDaveD** [comment, marked as answer]
Fix is in #321
